File: dsa-problems/Greedy/solutions/javascript/GRD-014-festival-bandwidth-split.js
const readline = require("readline");

class Solution {
  maxStages(n, B, bandwidths) {
    // Sort numerically!
    bandwidths.sort((a, b) => a - b);
    
    let currentSum = 0;
    let count = 0;
    
    for (const b of bandwidths) {
      if (currentSum + b <= B) {
        currentSum += b;
        count++;
      } else {
        break;
      }
    }
    
    return count;
  }
}

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

let data = [];
rl.on("line", (line) => data.push(line.trim()));
rl.on("close", () => {
  if (data.length === 0) return;
  
  let ptr = 0;
  const [n, B] = data[ptr++].split(" ").map(Number);
  const bandwidths = data[ptr++].split(" ").map(Number);

  const solution = new Solution();
  console.log(solution.maxStages(n, B, bandwidths));
});

File: dsa-problems/Greedy/solutions/python/GRD-014-festival-bandwidth-split.py
import sys

def max_stages(n: int, B: int, bandwidths: list) -> int:
    bandwidths.sort()
    
    current_sum = 0
    count = 0
    
    for b in bandwidths:
        if current_sum + b <= B:
            current_sum += b
            count += 1
        else:
            break
            
    return count

def main():
    input = sys.stdin.read
    data = input().split()
    if not data:
        return
        
    iterator = iter(data)
    n = int(next(iterator))
    B = int(next(iterator))
    
    bandwidths = []
    for _ in range(n):
        bandwidths.append(int(next(iterator)))

    result = max_stages(n, B, bandwidths)
    print(result)

if __name__ == "__main__":
    main()
